core/src/main/java/io/undertow/server/handlers/CurlRequestDumpingHandler.java

```diff
@@ -0,0 +1,114 @@
+        this.next = next;
+    }
+
+    public static boolean isPostOrPutMethod(HttpServerExchange exchange)
+    {
+        return (Methods.POST).equals(exchange.getRequestMethod()) || (Methods.PUT).equals(exchange.getRequestMethod());
+    }
+
+    @Override
+    public void handleRequest(final HttpServerExchange exchange) throws Exception
+    {
+
+        CURL_REQUEST_DUMP:
+        {
+            if (!UndertowLogger.REQUEST_DUMPER_LOGGER.isDebugEnabled() || !isPostOrPutMethod(exchange))
+            {
+                break CURL_REQUEST_DUMP;
+            }
+
+            final StringBuilder sb = new StringBuilder(2000);
+            sb.append("curl '" + exchange.getRequestScheme() + ":/" + exchange.getDestinationAddress()
+                    + exchange.getRequestURI());
+            String queryString = exchange.getQueryString();
+            if (exchange.getQueryString() != null && exchange.getQueryString().length() > 0)
+            {
+                sb.append("?" + queryString + "'");
+            }
+            else
+            {
+                sb.append("'");
+            }
+
+            Map<String, Cookie> cookies = exchange.getRequestCookies();
+            if (cookies != null)
+            {
+                for(Map.Entry<String, Cookie> entry : cookies.entrySet())
+                {
+                    Cookie cookie = entry.getValue();
+                    sb.append(" --cookie '" + cookie.getName() + "=" + cookie.getValue() + "'\n");
+                }
+            }
+            for(HeaderValues header : exchange.getRequestHeaders())
+            {
+                for(String value : header)
+                {
+                    sb.append(" -H '" + header.getHeaderName() + ": " + value + "'");
+                }
+            }
+
+            exchange.addRequestWrapper(new ConduitWrapper<StreamSourceConduit>()
+            {
+                @Override
+                public StreamSourceConduit wrap(final ConduitFactory<StreamSourceConduit> factory,
+                        final HttpServerExchange exchange)
+                {
+                    if (exchange.isRequestChannelAvailable() && !exchange.isResponseStarted())
+                    {
+                        ConduitStreamSourceChannel sourceChannel = ((HttpServerConnection) exchange.getConnection())
+                                .getChannel().getSourceChannel();
+                        dumpConduit = new CurlRequestDumpingConduit(sourceChannel.getConduit());
+                        return dumpConduit;
+                    }
+                    return factory.create();
+                }
+            });
+
+            exchange.addExchangeCompleteListener(new ExchangeCompletionListener()
+            {
+                @Override
+                public void exchangeEvent(final HttpServerExchange exchange, final NextListener nextListener)
+                {
+                    sb.append(" --data-binary '" + new String(dumpConduit.fullRequestDump()) + "'");

```

Comment:
This seems really problematic. In particular:
- What happens if the data starts with an @ character
- What happens if the data contains a ' character
- What happens if the data contains other non-ascii characters that will not be handled correctly?

I think you need to base64 encode this data and pipe it to curl. Something like:

echo '$base64encodeddata' | base64 | curl --data-binary @- ...

Rather than getting the binary data as a string you need to base64 encode the binary data. Of course if there is no data all this should be omitted and it should just call curl directly.